[ADD] module to manag promotion rules

bguillot opened the pull request.

bguillot committed:
[ADD] module to manag promotion rules

bguillot committed:
[IMP] add button to clean promotion rule

Cedric-Pigeon commented on sale_promotion_rule/models/sale_order.py:
```diff
@@ -0,0 +1,88 @@
+    @api.multi
+    def clear_promotion_line(self):
+        for order in self:
+            for line in order.order_line:
```
I would do it in a more compact way:
`order.order_line.filtered('promotion_rule_id')`

Cedric-Pigeon commented on sale_promotion_rule/models/sale_order.py:
```diff
@@ -0,0 +1,88 @@
+    def _can_be_discounted(self, rule):
+        self.ensure_one()
+        if self.discount and rule.use_best_discount \
+                and rule.discount_amount > self.discount:
```
I would use `float_compare` method

Cedric-Pigeon commented on sale_promotion_rule/models/sale_promotion_rule.py:
```diff
@@ -0,0 +1,134 @@
+            (self.date_from and fields.Date.today() < self.date_from))
+
+    def _check_valid_total_amount(self, order):
+        return self.minimal_amount < order[self.restriction_amount]
```
I would use `float_compare` method

Cedric-Pigeon commented on sale_promotion_rule/models/sale_promotion_rule.py:
```diff
@@ -0,0 +1,134 @@
+            if record.rule_type == 'coupon':
+                record.display_name = '%s (%s)' % (record.name, record.code)
+            elif record.rule_type == 'auto':
+                record.display_name = '%s (%s)' % (record.name, 'Automatic')
```
would be great to set 'Automatic' translatable

Cedric-Pigeon reviewed (changes requested):
Some small remarks/advices

lmignon committed:
[ADD] setup.py

lmignon commented on sale_promotion_rule/models/sale_order.py:
```diff
@@ -0,0 +1,88 @@
+    @api.multi
+    def clear_promotion(self):
+        for order in self:
+            order.promotion_rule_id = False
```
@bguillot An overall comment about your code. You should keep in mind that in Odoo each direct assignment to a field triggers a write into database and the methods registered on this field with `api.depends` ... I'll try to refactor your code to minimise the number of call to the database etc..

lmignon commented on sale_promotion_rule/models/sale_order.py:
```diff
@@ -0,0 +1,88 @@
+                    raise UserError(
+                        _('The rule cannot be applied on the sale order'))
+            else:
+                rules = self.env['sale.promotion.rule'].search([
```
@bguillot no search into a loop. Moreover, the rules could be searched before the loop since it's not based on item's information. (I'll provide a refactoring)

lmignon commented on sale_promotion_rule/models/sale_order.py:
```diff
@@ -0,0 +1,88 @@
+                    ('rule_type', '=', 'auto'),
+                ])
+                for rule in rules:
+                    if rule._is_promotion_valid(order):
```
@bguillot Very strange code.... You modify the current order to set the rule to apply... That means that at the end of the loop, the last rule applied is kept on the order...

lmignon committed:
[IMP] sale_promotion_rule: More batch processes

lmignon committed:
[PYLINT] sale_promotion_rule

lmignon committed:
[IMP] sale_promotion_rule: Improve implementation

The module now supports multiple promotion rules.
When a coupon is used, it always take precedence on automatic promotion
rule.
The computation of promotions has been refactored to update lines from
the sale.order object. Sine lines are updated from the SO, the recompute
of SO fields that depends of lines is only tiggered at the end of the
update on the lines. (In place of each time a line is updated).
You can ask to recompute the promotion of a sale.order. If a coupon is
already specified it's conserved but all the automatic rules are applied
again

lmignon committed:
[IMP] sale_promotion_rule: Wizard to add coupon + improved UI

lmignon committed:
[FIX] typo

lmignon committed:
[IMP] sale_promotion_rule: Avoid nested view definition

lmignon committed:
[FIX] sale_promotion_rule: _apply must support multi recordsed

lmignon committed:
[FIX] sale_promotion_rule: remove unsed code

sebastienbeau committed:
Merge pull request #116 from acsone/10.0-sale_promotion_rule-improved-lmi

10.0 sale promotion rule improved lmi

lmignon commented:
Thank you @bguillot and @sebastienbeau
